site/src/hooks/useClipboard.test-setup.ts

```diff
@@ -0,0 +1,172 @@
+ * the test file as the main boundary for test environments, with each test case
+ * able to run in parallel. That works if you have one single global state, but
+ * we need two separate versions of the global state, while repeating the exact
+ * same test cases for each one.
+ *
+ * If both tests were to be placed in the same file, Jest would not isolate them
+ * and would let their setup steps interfere with each other. This leads to one
+ * of two things:
+ * 1. One of the global mocks overrides the other, making it so that one
+ *    connection type always fails
+ * 2. The two just happen not to conflict each other, through some convoluted
+ *    order of operations involving closure, but you have no idea why the code
+ *    is working, and it's impossible to debug.
+ */
+import { type UseClipboardResult, useClipboard } from "./useClipboard";
+import { act, renderHook } from "@testing-library/react";
+
+const initialExecCommand = global.document.execCommand;
+beforeAll(() => {
+  jest.useFakeTimers();
+});
+
+afterAll(() => {
+  jest.restoreAllMocks();
+  jest.useRealTimers();
+  global.document.execCommand = initialExecCommand;
+});
+
+type MockClipboardEscapeHatches = Readonly<{
+  getMockText: () => string;
+  setMockText: (newText: string) => void;
+}>;
+
+type MockClipboard = Readonly<Clipboard & MockClipboardEscapeHatches>;
+function makeMockClipboard(isSecureContext: boolean): MockClipboard {
+  let mockClipboardValue = "";
+
+  return {
+    readText: async () => {
+      if (!isSecureContext) {
+        throw new Error(
+          "Trying to read from clipboard outside secure context!",
+        );
+      }
+
+      return mockClipboardValue;
+    },
+    writeText: async (newText) => {
+      if (!isSecureContext) {
+        throw new Error("Trying to write to clipboard outside secure context!");
+      }
+
+      mockClipboardValue = newText;
+    },
+
+    getMockText: () => mockClipboardValue,
+    setMockText: (newText) => {
+      mockClipboardValue = newText;
+    },
+
+    addEventListener: jest.fn(),
+    removeEventListener: jest.fn(),
+    dispatchEvent: jest.fn(),
+    read: jest.fn(),
+    write: jest.fn(),
+  };
+}
+
+function renderUseClipboard(textToCopy: string) {
+  return renderHook<UseClipboardResult, { hookText: string }>(
+    ({ hookText }) => useClipboard(hookText),
+    { initialProps: { hookText: textToCopy } },
+  );
+}
+
+type ScheduleConfig = Readonly<{ isHttps: boolean }>;
+
+export function scheduleClipboardTests({ isHttps }: ScheduleConfig) {
+  const mockClipboardInstance = makeMockClipboard(isHttps);
+
+  beforeAll(() => {
+    const originalNavigator = window.navigator;
+    jest.spyOn(window, "navigator", "get").mockImplementation(() => ({
+      ...originalNavigator,
+      clipboard: mockClipboardInstance,
+    }));
+
+    if (!isHttps) {
+      // Not the biggest fan of exposing implementation details like this, but
+      // making any kind of mock for execCommand is really gnarly in general
```

Comment:
Very neat you were able to test the clipboard with a unit test!  It is unfortunate we have to mock the clipboard and `execCommand` though, I feel like if there is a bug there is a big chance it will be something about our use of these APIs so they will not be caught unless the mock is perfect.  But this is good I think, maybe it is something we can look into later with the Storybook actions thing since I think that gives us a real browser?  Or maybe it would have to be full-on e2e.

Reply:
Yeah, I do think that an E2E test could be more reliable (and I'm planning to start really learning Playwright soon, so maybe a test rewrite could be done sooner rather than later), but I'm not sure if they give you a way to set the security context

Maybe something like the [`APIRequest`](https://playwright.dev/docs/api/class-apirequest#api-request-new-context) API could do the trick, though?

Reply:
Oooh interesting, I was not aware of `APIRequest`.  In code-server we spawn twice, one on http and one on https, but it is pretty unfortunate to do it that way.